File: HookHook.Web/src/store/service.ts
import { authHeader } from ".";

const service = {
    namespaced: true,
    state: {
        accounts: {}
    },
    mutations: {
        getAccounts(state: any, {provider, info }: any) {
            state.accounts[provider] = info;
        },
        deleteAccount(state: any, { provider, key }: any) {
            state.accounts[provider].splice(key, 1);
        },
        addAccount(state: any, {provider, info}: any) {
            state.accounts[provider].push(info);
        }
    },
    actions: {
        async getAccounts({ commit, state }: any, provider: any) {
            const res = await fetch('/api/service/' + provider, {
                method: 'GET',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                commit('getAccounts', { provider, info: json });
            }
            return {};
        },
        async deleteAccount({ commit }: any, { provider, id, key }: any) {
            const res = await fetch('/api/service/' + provider + '?id=' + id, {
                method: 'DELETE',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                return await res.json();
            }
            commit('deleteAccount', { provider, key });
            return {};
        },
        async addDiscord({ commit }: any, code: String) {
            const res = await fetch("/api/service/discord?code=" + code, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'Discord', info: json });
            }
            return {};
        },
        async addTwitter({ commit }: any, { token, verifier }: any) {
            const res = await fetch(`/api/service/twitter?code=${token}&verifier=${verifier}`, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'Twitter', info: json });
            }
            return {};
        },
        async addTwitch({ commit }: any, code: String) {
            const res = await fetch("/api/service/twitch?code=" + code, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'Twitch', info: json });
            }
            return {};
        },
        async addSpotify({ commit }: any, code: String) {
            const res = await fetch("/api/service/spotify?code=" + code, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'Spotify', info: json });
            }
            return {};
        },
        async addGitHub({ commit }: any, code: String) {
            const res = await fetch("/api/service/github?code=" + code, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'GitHub', info: json });
            }
            return {};
        },
        async addGoogle({ commit }: any, code: String) {
            const res = await fetch("/api/service/google?code=" + code, {
                method: 'POST',
                headers: authHeader()
            });
            if (res.status === 500) {
                return { error: "Backend unavailable" };
            }
            const contentType = res.headers.get("content-type");
            if (contentType && (contentType.indexOf("application/json") !== -1 || contentType.indexOf("application/problem+json") !== -1)) {
                const json = await res.json();
                if (json.error !== undefined || json.errors !== undefined)
                    return json;
                if (json.userId === undefined || json.username === undefined) {
                        return {};
                }
                commit('addAccount', { provider: 'Google', info: json });
            }
            return {};
        },
    }
}

export default service

File: HookHook.Mobile/lib/wrapper/service.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';

import '../main.dart';
import 'backend.dart';
import 'package:http/http.dart' as http;

class Account {
  Account({
    required this.userId,
    required this.username,
  });

  final String userId;
  final String username;

  factory Account.fromJson(Map<String, dynamic> json) => Account(
    username: json["username"],
    userId: json["userId"],
  );

  Map<String, dynamic> toJson() => {
    "userId": userId,
    "username": username,
  };
}

class Service {
  static String baseUrl = Backend.apiEndpoint + "service/";
  static String twitterUrl = baseUrl + "twitter";
  static String twitchUrl = baseUrl + "twitch";
  static String spotifyUrl = baseUrl + "spotify";
  static String googleUrl = baseUrl + "google";
  static String gitHubUrl = baseUrl + "github";
  static String discordUrl = baseUrl + "discord";

  Future<List<Account>> getAccounts(String provider) async {
    final res = await http.get(Uri.parse("$baseUrl$provider"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    ).timeout(const Duration(
        seconds: 3
    ));
    if (res.statusCode == 200) {
      return List<Account>.from(
          jsonDecode(res.body).map((x) => Account.fromJson(x)));
    }
    return [];
  }

  Future<void> deleteAccount(String provider, String id) async {
    final res = await http.delete(Uri.parse("$baseUrl$provider?id=$id"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    ).timeout(const Duration(
        seconds: 3
    ));
    if (res.statusCode != 204) {
      if (kDebugMode) {
        print("DELETE ACCOUNT FAILED");
      }
    }
  }

  Future<Account?> addTwitter(String code, String verifier) async {
    final res = await http.post(Uri.parse(
        "$twitterUrl?code=$code&verifier=$verifier"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }

  Future<Account?> addTwitch(String code) async {
    final res = await http.post(Uri.parse(
        "$twitchUrl?code=$code"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }

  Future<Account?> addSpotify(String code) async {
    final res = await http.post(Uri.parse(
        "$spotifyUrl?code=$code&redirect=hookhook://oauth/spotify"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }

  Future<Account?> addGoogle(String code) async {
    final res = await http.post(Uri.parse(
        "$googleUrl?code=$code"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }

  Future<Account?> addGitHub(String code) async {
    final res = await http.post(Uri.parse(
        "$gitHubUrl?code=$code"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }

  Future<Account?> addDiscord(String code, String verifier) async {
    final res = await http.post(Uri.parse(
        "$discordUrl?code=$code&verifier=$verifier&redirect=hookhook://oauth/discord"),
      headers: {
        HttpHeaders.authorizationHeader: "Bearer " +
            (await HookHook.backend.signIn.token)!,
      },
    )
        .timeout(
        const Duration(
            seconds: 3
        )
    );
    if (res.statusCode == 200) {
      return Account.fromJson(jsonDecode(res.body));
    }
    return null;
  }
}